scripts/tooling/vuart.c

```diff
@@ -184,6 +187,187 @@ struct tenstorrent_configure_tlb {
+{
+	int fd;
+	int ret;
+
+	/*
+	 * easier to use open / read / strtoul in this case than fopen / fscanf because the
+	 * former supports optional 0x prefix and automatically detects base.
+	 */
+
+	fd = open(path, O_RDONLY);
+	if (fd < 0) {
+		D(1, "Failed to open %s: %s", path, strerror(errno));
+		return 0;
+	}
+
+	ret = read(fd, path, MIN(path_size, 32));
+	if (ret < 0) {
+		E("Failed to read from %s: %s", path, strerror(errno));
+		goto out;
+	}
+
+	path[ret] = '\0';
+	errno = 0;
+	*value = strtoul(path, NULL, 0);
+	ret = -errno;
+
+out:
+	close(fd);
+
+	return ret;
+}
+
+static int pcie_walk_sysfs(uint16_t match_vid, uint16_t match_pid,
+			   int (*cb)(char *path, size_t path_size, void *data), void *data)
+{
+	int ret;
+	DIR *dir;
+	unsigned long pid;
+	unsigned long vid;
+	struct dirent *dent;
+	static char path[1024];
+	unsigned int counter = 0;
+
+	strncpy(path, PCI_DEVICES_PATH, sizeof(path));
+	dir = opendir(path);
+	if (dir == NULL) {
+		E("Failed to open %s: %s", PCI_DEVICES_PATH, strerror(errno));
+		return -errno;
+	}
+
+	for (dent = readdir(dir); dent != NULL; dent = readdir(dir)) {
+
+#ifndef DT_LNK
+#define DT_LNK 10
+#endif
+
+		D(3, "Found %s/%s with d_type %d", PCI_DEVICES_PATH, dent->d_name, dent->d_type);
+
+		if (dent->d_type == DT_LNK) {
+			snprintf(path, sizeof(path), "%s/%s/vendor", PCI_DEVICES_PATH,
+				 dent->d_name);
+			ret = pcie_read_unsigned_long_from_file(path, sizeof(path), &vid);
+			if (ret != 0) {
+				continue;
+			}
+			snprintf(path, sizeof(path), "%s/%s/device", PCI_DEVICES_PATH,
+				 dent->d_name);
+			ret = pcie_read_unsigned_long_from_file(path, sizeof(path), &pid);
+			if (ret != 0) {
+				continue;
+			}
+
+			if (((vid == match_vid) || (match_vid == 0xffff)) &&
+			    ((pid == match_pid) || (match_pid == 0xffff))) {
+
+				D(1, "Found %s/%s with vendor id %04lx and product id %04lx",
+				  PCI_DEVICES_PATH, dent->d_name, vid, pid);
+
+				snprintf(path, sizeof(path), "%s/%s", PCI_DEVICES_PATH,
+					 dent->d_name);
+
+				ret = cb(path, sizeof(path), data);
+				if (ret < 0) {
+					goto out;
+				}
+				counter += ret;
+			}
+		}
+	}
+
+	ret = counter;
+
+out:
+	closedir(dir);
+
+	return ret;
+}
+
+static int pcie_remove_cb(char *path, size_t path_size, void *data)
+{
+	int fd;
+	int ret;
+
+	(void)data;
+
+	strncat(path, "/remove", path_size);
+
+	fd = open(path, O_WRONLY);
+	if (fd < 0) {
+		E("Failed to open %s: %s", path, strerror(errno));
+		return -errno;
+	}
+
+	if (write(fd, "1", 1) < 0) {
+		E("Failed to write to %s: %s", path, strerror(errno));
+		ret = -errno;
+		goto out;
+	}
+
+	path[strlen(path) - 7] = '\0';
+	D(1, "Removed PCIe device %s", path);
+
+	ret = 1; /* success: count the number of devices removed */
+
+out:
+	close(fd);
+
+	return ret;
+}
+
+int pcie_remove(void)
+{
+	int ret;
+
+	ret = pcie_walk_sysfs(TENSTORRENT_PCI_VENDOR_ID, 0xffff, pcie_remove_cb, NULL);
+
+	if (ret > 0) {
+		D(1, "Removed %d PCIe devices", ret);
+	}
+
+	return ret;
```

Comment:
This could be simplified

```suggestion
int pcie_remove(void)
{
	return pcie_walk_sysfs(TENSTORRENT_PCI_VENDOR_ID, 0xffff, pcie_remove_cb, NULL);
```